Feature/agent add turris omnia rdkb support

morantr opened the pull request:
Port Agent changes from the Turris Omnia support branch (#307).

tomereli commented on agent/CMakeLists.txt:
```diff
@@ -7,37 +7,40 @@ if(NOT CMAKE_BUILD_TYPE)
 message(STATUS "Beerocks Build Type: ${CMAKE_BUILD_TYPE}")
 
+# Default values
+    set(BEEROCKS_WLAN1_IFACE "wlan0")
```
Wrong indentation. Also - @vitalybu , @arnout - isn't this better if we save the interface names in the platform DB instead of here? Then we can completely remove Turris Omnia with generic OpenWRT.

morantr replied:
I'll fix the indentation, but to your question about the interfaces, the answer is no.
The interfaces names are set in the configuration files and are needed before we have the ability to get them from bpl.
For example, the son_slave_main needs the interfaces from the configuration to decide which radio son_slave to bring up. I guess we can find a way to remove completely the configuration file and move everything to the database, but we need to make DR for it.

tomereli replied:
ok.

tomereli commented on agent/CMakeLists.txt:
```diff
@@ -7,37 +7,40 @@ if(NOT CMAKE_BUILD_TYPE)
-    set(BEEROCKS_REPEATER_MODE 0)
-    set(BEEROCKS_AUTOMATIC_CREDENTIALS_UNIFY 0)
-    set(BEEROCKS_STOP_ON_FAILURE_ATTEMPTS 1)
+elseif (TARGET_PLATFORM STREQUAL "turris-omnia")
```
Again - we shouldn't have turris-omnia target, just openwrt target.

vitalybu replied:
I'll change it as part of the UGW porting PR

tomereli commented on agent/CMakeLists.txt:
```diff
@@ -108,6 +111,12 @@ add_definitions(-DBEEROCKS_REVISION="${${PROJECT}_REVISION}")
 if(TARGET_PLATFORM STREQUAL "linux")
 add_definitions(-DBEEROCKS_BIN_PATH="${CMAKE_INSTALL_PREFIX}/bin/")
 add_definitions(-DBEEROCKS_CONF_PATH="${CMAKE_INSTALL_PREFIX}/config/")
+elseif(TARGET_PLATFORM STREQUAL "turris-omnia")
```
I don't like this. We should use the same paths on all platforms. no more `/opt/beerocks` please.

vitalybu replied:
I'll change it as part of the UGW porting PR

tomereli commented on common/beerocks/tlvf/CMakeLists.txt:
```diff
@@ -3,7 +3,13 @@ cmake_minimum_required (VERSION 2.8.12)
 project(${PROJECT_NAME})
 
-find_package(PythonInterp REQUIRED)
+# TODO: FIXME: For some reason CMake detects only the Python from the cross-toolchain
```
I suspect that this happened in the demo preparations. Now is the time for fixing this - don't leave it as TODO. Same goes for all other CMAKE hacks (but I'll comment on those separately).

tomereli replied:
Also - What is the problem with the python from the cross-toolchain? Why can't we just use it?

vitalybu replied:
The problem is with the pyYAML library dependency. Not all cross-toolchains contain this package.
As for this specific hack, it may be an issue with the older Turris tree (3.x), perhaps it doesn't exist in the latest one.
I suggest opening an issue and handling it separably from this PR.

tomereli replied:
Can't we quickly check if this is needed in the new Turris tree? I think @arnout or @rmelotte can help with a quick check.
Of the answer is that it is needed for now, then @morantr please open an issue on that and add the issue link in the TODO comment.

arnout replied:
I've reverted this bit on the dev/turris-omnia-support branch and it built fine.

Anyway it doesn't make any sense that this would be needed in common but not in framework.

Therefore, please don't include this change.

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -616,13 +616,15 @@ bool main_thread::backhaul_fsm_main(bool &skip_select)
             FSM_MOVE_STATE(MASTER_DISCOVERY);
         } else { // link establish
 
+#ifdef BEEROCKS_LINUX
```
This commit doesn't explain *why* this is done and also why is it done only for Linux.
The commit message should be something like "agent: backhaul: skip removing STA interfaces from the bridge" with a description which explains why.

In addition - the *only* mode is passive mode so we should *never* do any bridge operations. This commit should remove this logic completely.

morantr replied:
Add a fix up that remove it

tomereli commented on agent/src/beerocks/slave/platform_manager/platform_manager_thread.cpp:
```diff
@@ -1518,7 +1518,7 @@ bool main_thread::handle_cmdu(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx)
             m_pBackhaulManagerSlave = sd;
 
-#if !defined(BEEROCKS_LINUX) && !defined(BEEROCKS_TURRIS_OMNIA)
+#ifndef BEEROCKS_TURRIS_OMNIA
```
Whole commit is wrong. Doesn't say why it disables arp monitor on Turris Omnia, and also it Re-enables arp monitor for Linux.
What was wrong with how it used to be? Looks like this commit should be completely removed.

vitalybu replied:
Turris Omnia currently uses the same dummy BPL and the Linux platform, so the ARP monitor should be disabled for both. Later on we should create a Linux BPL with ARP monitor support and use it for all Linux based platforms.

tomereli replied:
But now arp monitor will be enabled for Linux, is this what you meant to happen?

Does the arp monitor use the BPL? Is this why it has to be disabled when dummy BPL is used?

vitalybu replied:
No, enabling the ARP monitor for Linux is a bug.
It should be disabled for both Linux and Turris Omnia platforms.

tomereli replied:
Still not fixed.

morantr replied:
removing the commit

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -581,16 +581,6 @@ bool main_thread::backhaul_fsm_main(bool &skip_select)
             FSM_MOVE_STATE(MASTER_DISCOVERY);
         } else { // link establish
 
-#ifdef BEEROCKS_LINUX
```
This is fine but don't forget to fix the commit message once you rebase

tomereli reviewed (approved):
This is a critical path change so I'm already approving - but please fix the bug you added that enables arp monitor for linux, and address the remaining comments. After that, you can merge IMO.

morantr committed:
agent: Adjust the agent to support Turris Omnia platform

cmake adjustments to compile and run successfully prplMesh on Turris
Omnia platform.

Changes include:
- Create default cmake macros which are common to most of the platforms.
- Setting platform paths.
- Explicitly link with MapfTransport.
When building for OpenWrt, for some reason it doesn't find MapfTransport
which is a transitive dependency from btl. So add it explicitly to the
link of agent and controller for the time being.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
agent: Fix bug which caused to packets to be sampled incorrectly

This commit porting rdkb monitor changes.
Packets were sampled incorrectly when inactCheckIntervalSec is set to
a value greater than 1.
Fix bug that caused to threshold update to occur only on multiply of
interval  interval_sec.
Also, fix the wrong presentation of packets counter on rdkb monitor.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
agent: Add generation of connected clients events

On passive mode, clients may be able to connect to the Agent before it
has a link to the controller.
Add a client association notification generation request which will
simulate client association notification events on clients that are
already connected.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
agent: Add to backhaul manager pointer validation and prints

Add to backhaul manager pointer validation and prints, and fix wrong
pointers resets.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
agent: Add backhaul manager reset if its platform manager disconnect

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
agent: Skip removing STA interfaces from the bridge

Since the only mode is passive mode so we should never do any bridge
operations.
Remove the piece of code which removes the station interfaces from the
bridge.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
agent: Fix the wrong comparison to slave interface on backhaul manager

Replace comparison of radio interface name to a constant interface name
such as "wlan0"  Freq type, since interfaces name may be different on
different platforms.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
agent: Move to MASTER_DISCOVERY state instead of INITIATE_SCAN

When the backhaul manager perform reset for any reason, and re-attach
to bwl, the backhaul may be already connected. In that case, there is no
point to perform the whole connecting sequence, so jump immediately to
MASTER_DISCOVERY state.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
agent: Enable backhaul preferred radio band feature

Add feature which allows connecting the backhaul radio on a preferred
band, which is configured on the platform database.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
agent: Modify backhaul manager WIRELESS_WAIT_FOR_RECONNECT state

Change the state WIRELESS_WAIT_FOR_RECONNECT on backhaul manager to
enter the AP into the blacklist and move to INITIATE_SCAN mode instead
of WIRELESS_ASSOCIATE_4ADDR state, since the AP might no longer
transmitting. This change allows removing "wifi_reconnect_flag" which
is no longer useful.

Since this change affect on WIRELESS_ASSOCIATE_4ADDR state which now
only consider the roam_flag, so I add a comment on using the roaming
flow that way.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
agent: Add AP enable flow

Add AP enable flow which is used on passive repeater mode.
In this mode, the backhaul manager sends upon backhaul connection to
the AP manager a request to enable the AP, which means it will start
beaconing. After the operation will be completed, the AP manager
will send an "AP-DISABLED" event and the son slave will restart.
The backhaul manager will perform a reset too but since it has already
a backhaul connection it will skip link establishment flow and
immediately do the discovery sequence.

This commit includes:
- all messages yaml, and auto-generated file which is
used in this flow.
- Add a function that sends the AP_ENABLE request on the backhaul
manager on the backhaul connection.
- Add an AP_ENABLE message handler on ap manager.
- Add message forward handler for AP_ENABLE request, and handler for
the response.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
agent: Remove DHCP flows from backhaul manager

These flows are not being used on passive mode since the platform is
responsible for getting an IP address.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Fix bug which caused to move to the wrong backhaul state

Replace the "break" statement to "return" since the break statement will
break the loop and no  the switch case, which will allow changing the
state after the loop.
In addition, remove the
`if (FSM_IS_IN_STATE(WIRELESS_ASSOCIATE_4ADDR_WAIT)) {` which wrap it
since we are already in this state so it is useless the check it again.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
agent: Add to monitor handler for beacon request status

Add to monitor handler for beacon request status for updating the dialog
token on pending requests.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
agent: Backhaul manager socket reuse if exist

Instead of removing backhaul manager socket on son_slave
STATE_CONNECT_TO_BACKHAUL_MANAGER and create new one,
use it if it already open.

Also fix/add some debug prints.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
agent: Add 'mem_only_psk_flag' to BACKHAUL_ENABLE message

'mem_only_psk' flag sets whether the platform shall save the credentials
in some encrypted DB so the wpa-supplicant will get from there in
runtime, without supply it on `connect` API, or not.

Add the flag to the BACKHAUL_ENABLE yaml and auto generated files.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

arnout commented on agent/src/beerocks/slave/CMakeLists.txt:
```diff
@@ -34,6 +35,11 @@ add_executable(${PROJECT_NAME} ${beerocks_agent_sources})
 set_target_properties(${PROJECT_NAME} PROPERTIES LINK_FLAGS "-Wl,-z,defs")
 target_link_libraries(${PROJECT_NAME} rt dl beerocks::bcl beerocks::btl tlvf elpp beerocks::btlvf ${LIBS})
 
+# For some reason, this isn't found automatically from btl...
```
I'll try to fix this once top-level CMake is merged.

arnout reviewed (changes requested):
Reviewed only the first 3 commits for now. I'm already submitting this review so the find_package(PythonInterp) change can be undone. Feel free to rebase and force-push after fixing.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -589,18 +589,19 @@ bool main_thread::backhaul_fsm_main(bool &skip_select)
-            m_scPlatform.reset();
-        }
-
         m_scPlatform = std::make_shared<SocketClient>(
```
Indentation is wrong here after commit "agent: Add to backhaul manager pointer validation and prints" (but fixed later). (nitpick)

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -589,18 +589,19 @@ bool main_thread::backhaul_fsm_main(bool &skip_select)
         }
+        } else {
+            LOG(DEBUG) << "Using existing platform_manager_socket=0x"
+                       << intptr_t(m_scPlatform.get());
```
AFAIU this will print decimal, not hex (intptr_t is an integer type, not a pointer type, so default representation is decimal). (nitpick)

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -690,18 +690,15 @@ bool main_thread::backhaul_fsm_main(bool &skip_select)
+
         for (auto soc : slaves_sockets) {
-            if (soc->hostap_iface == "wlan0") {
+            if (soc->freq_type == eFreqType::FREQ_24G) {
```
This code is still incorrect for a 3-radio system (2.4G, 5GL, 5GH). But that can be fixed later.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1010,6 +1023,14 @@ bool main_thread::backhaul_fsm_wireless(bool &skip_select)
+            FSM_MOVE_STATE(MASTER_DISCOVERY);
+            state_attempts = 0; // for next state
+        } else {
+            FSM_MOVE_STATE(INITIATE_SCAN);
```
I actually don't think prplMesh should manage the wireless backhaul link at all (where "manage" means: configure it; it should still monitor the backhaul link, and trigger roaming if the controller requests it). The platform should already have brought the backhaul link up before prplMesh starts (though of course it's possible that the link is not up yet, therefore monitoring is needed). And wpa_supplicant is perfectly capable of selecting a BSS all by itself.

The only reasons for prplMesh to be involved is for channel selection, and for roaming between bands which would require a switch of radio. For channel selection, I think it's up to the controller to make such a decision - it can take into account the channel the backhaul station will force the radio to after steering to a different BSS.

Roaming between radios is more tricky, there is unfortunately nothing in EasyMesh that stops a controller from doing that. It's a tricky thing because you generally don't want to bring down the old STA before the new one is up and running. So I wouldn't try to deal with that for now. And anyway, I think it should be handled by having the backhaul STA config set up for both radios, but leaving one of them disabled.

Note that in the latter context, the "backhaul preferred radio band feature" *is* relevant and should be kept. So +1 for that change.

Anyway, all of this is not for this PR, just an observation.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1140,63 +1140,57 @@ bool main_thread::backhaul_fsm_wireless(bool &skip_select)
         if (roam_flag) {
             selected_bssid         = roam_selected_bssid;
             selected_bssid_channel = roam_selected_bssid_channel;
-        }
-
-        if (wifi_reconnect_flag) {
             if (!active_hal->roam(selected_bssid, selected_bssid_channel)) {
```
Before the "Modify backhaul manager WIRELESS_WAIT_FOR_RECONNECT state", no roaming was actually done when roam_flag was set (or rather, it was only done after a disconnect event). Now, roaming is done immediately. I assume that that is intentional - but it's not so clear from the commit message. Could you double check that this roam call should be done when roam_flag is set, and should not be removed instead?

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1535,6 +1539,15 @@ bool main_thread::handle_slave_backhaul_message(std::shared_ptr<SSlaveSockets> s
+        * Another thing it is causing is to re-scan the for networks. Is it really necessary?
+        * Can we just move immediately to WIRELESS_ASSOCIATE_4ADDR, and then when successfully
+        * connected, notify the agent about the ne bssid?
+        */
```
Great that you added this explanation!

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1517,7 +1518,7 @@ bool main_thread::handle_slave_backhaul_message(std::shared_ptr<SSlaveSockets> s
                     m_sConfig.wired_backhaul  = request->wired_backhaul();
                     LOG(DEBUG) << "All slaves ready, proceeding" << std::endl
-                               << "SSID: " << m_sConfig.ssid << ", Pass: " << m_sConfig.pass
+                               << "SSID: " << m_sConfig.ssid << ", Pass: ****"
```
I'm not sure why to use *** here. The credentials will be readable in the hostapd config file anyway.

If you're removing the password, just remove it.

(nitpick).

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1922,9 +1951,11 @@ bool main_thread::hal_event_handler(bwl::base_wlan_hal::hal_event_ptr_t event_pt
+            // Send slaves to enable the AP's
+            send_slaves_enable();
+
+            FSM_MOVE_STATE(MASTER_DISCOVERY);
```
This change is done as part of "Add AP enable flow" but I don't see how it's related. I think it should be part of "Remove DHCP flows".

BTW, why is the condition on WIRELESS_ASSOICATE_4ADDR_WAIT removed? I think this condition is there to handle the cases where the state has changed in the loop above. If you now unconditionally change the state, those state changes above are useless. Oh, I see this is fixed in the commit "agent: Fix bug which caused to move to the wrong backhaul state" - so perhaps that should be squashed into this one (or rather, in "Remove DHCP flows" which should contain this change as well).

(nitpick)

morantr replied:
No, it should be done as part of this commit, because we must go to discovery when enabling the APs.
The WIRELESS_ASSOICATE_4ADDR_WAIT condition was removed because we already check it before (see line 1905).

arnout commented on agent/src/beerocks/slave/ap_manager_thread.cpp:
```diff
@@ -441,6 +441,70 @@ bool ap_manager_thread::handle_cmdu(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_
+
+        response->success() = true;
+
+        // Enable beaconing by setting start_disabled flag to false.
+        if (!ap_wlan_hal->set_start_disabled(false)) {
+            LOG(ERROR) << "Failed setting start_disabled";
+            response->success() = false;
+            message_com::send_cmdu(slave_socket, cmdu_tx);
+            break;
+        }
+
+        // Disable the radio interface to make hostapd to consider the new configuration.
+        if (!ap_wlan_hal->disable()) {
+            LOG(ERROR) << "Failed disable";
+            response->success() = false;
+            message_com::send_cmdu(slave_socket, cmdu_tx);
+            break;
+        }
+
+        // If it is not the radio of the BH, then channel, bandwidth and center channel paramenters
+        // will be all set to 0.
+        LOG(DEBUG) << "Setting AP channel: "
+                   << ", channel=" << int(notification->channel())
+                   << ", bandwidth=" << int(notification->bandwidth())
+                   << ", center_channel=" << int(notification->center_channel());
+
+        // Set original channel or BH channel
+        if (!ap_wlan_hal->set_channel(notification->channel(), notification->bandwidth(),
+                                      notification->center_channel())) {
+            LOG(ERROR) << "Failed setting set_channel";
+            response->success() = false;
+            message_com::send_cmdu(slave_socket, cmdu_tx);
+            break;
+        }
+
+        // Enable the radio interface to apply the new configuration.
+        if (!ap_wlan_hal->enable()) {
+            LOG(ERROR) << "Failed enable";
+            response->success() = false;
+        }
```
As far as I can see, the only thing that we do here is to set the channel, and to enable the AP if it wasn't enabled yet. I think the disable, if it is needed, should be handled by bwl itself instead of by ap_manager, because a good implementation should be able to change the channel without disabling the AP (and it should also be able to tell clients to roam to the new channel).

And of course, I think all of this shouldn't be needed because we shouldn't manage the backhaul from prplMesh, but that's a different issue.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1500,6 +1500,7 @@ bool main_thread::handle_slave_backhaul_message(std::shared_ptr<SSlaveSockets> s
                     m_sConfig.ssid.assign(request->ssid(message::WIFI_SSID_MAX_LENGTH));
                     m_sConfig.pass.assign(request->pass(message::WIFI_PASS_MAX_LENGTH));
                     m_sConfig.security_type = static_cast<bwl::WiFiSec>(request->security_type());
+                    m_sConfig.mem_only_psk  = request->mem_only_psk();
```
This mem_only_psk change belongs to a different commit (nitpick).

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1833,11 +1833,9 @@ bool main_thread::hal_event_handler(bwl::base_wlan_hal::hal_event_ptr_t event_pt
                             LOG(DEBUG) << "iface " << iface
                                        << " is connected on low 5G band with filter, aborting";
-                            if (FSM_IS_IN_STATE(WIRELESS_ASSOCIATE_4ADDR_WAIT)) {
                                 FSM_MOVE_STATE(WIRELESS_CONFIG_4ADDR_MODE);
```
Indentation is wrong in commit "agent: Fix bug which caused to move to the wrong backhaul state", fixed in a later commit. Same a few lines below. (nitpick)

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3249,23 +3249,25 @@ bool slave_thread::slave_fsm(bool &call_slave_select)
+                break;
         } else {
             add_socket(backhaul_manager_socket);
+            }
```
Indentation is wrong in "agent: Backhaul manager socket reuse if exist" commit, fixed later. (nitpick)

arnout reviewed (approved):
Lots of comments but mostly nitpicks or bigger issues.

I want the find_package(PythonInterp) change to be removed before merging, but otherwise good to go. The nitpicks are mostly about wrong commit splitting so not so important.

We should have a discussion about how to handle wireless backhaul. I made some notes in my review comments here, but it warrants a bigger discussion.
